security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -394,10 +390,9 @@ func (s *sdsservice) clearStaledClients() {
 		sdsServiceLog.Debugf("remove staled clients %+v", connKey)
 		delete(sdsClients, connKey)
 		delete(staledClientKeys, connKey)
+		// totalStaleConnCounts should be 0 when the for loop finishes.
+		totalStaleConnCounts.Decrement()
```

Comment:
Could you add a check and print error if totalStaleConnCounts is not equal to 0, and repeatedly call Decrement() to make it 0?

Reply:
or does it support reset to 0 directly (to avoid decrement in each loop) ?

Reply:
I can add some log. In theory, it should be set to 0 correctly (from my PR to make sure we only delete each key at most once, I also manually tested it). 

There is no support for setting it to 0 directly. I can add some code to the monitoring package to do this if you guys think it's necessary.

Reply:
Let's wait for https://github.com/istio/pkg/pull/35

Reply:
Actually, `RetrieveData` is intended for testing only https://github.com/census-instrumentation/opencensus-go/blob/master/stats/view/worker.go#L97, so probably we don't have a good way to access the metric value. However, the `totalStaleConnCounts` should be 0 since we don't add and delete the same keys twice. I also tested this.

Reply:
@quanjielin @JimmyCYJ Any other comments?

Reply:
Could you add test to verify that totalStaleConnCounts is 0 after clearStaledClients?

Reply:
This https://github.com/istio/istio/pull/15506 should already ensure we recycle each connection once.

Reply:
That is not covered by test either. It is better to have a test, so that any changes in the future that affect the metric would be caught by test.

Reply:
Sg. Will add some tests :)

Reply:
I updated all unit tests in `sdsservice_test.go` to check for number of staled connections by @JimmyCYJ's request. In order to do that, I changed `clearStaledClients()` to a function instead of a member method. This makes more sense since `clearStaledClients()` doesn't use any variables associated with the struct.